### DISCUSSION TITLE
Using BT for BadUSB-Style Input

### DISCUSSION THREAD

**Discussion by SimplyMinimal** [akopachov/flipper-zero_authenticator#81] [Ideas]
This is a brilliant idea to utilize BadUSB to enter the code ([reference](https://github.com/akopachov/flipper-zero_authenticator/wiki/Manually-typing-codes-is-boring%2C-is-there-a-Bad-USB-mode%3F)).

Looks like there's an existing implementation of using BadUSB over Bluetooth. Are there any plans to incorporate a BT version for a fully-wireless set up?

https://github.com/ClaraCrazy/Flipper-Xtreme/tree/main/applications/main/bad_kb

**Comment by akopachov** [comment]
Nice idea! Makes sense at least to give it a try to see how much useful it could be. Thank you.

**Reply by akopachov** [reply]
Logged [project item](https://github.com/users/akopachov/projects/1?pane=issue&itemId=20920253)

**Reply by eried** [reply]
+1!!!! this would be brilliant!

**Reply by akopachov** [reply]
Work is in progress!

**Reply by eried** [reply]
excellent! this is becoming the only app that makes me carry the flipper all the time :P not just for the lolz

**Reply by akopachov** [reply]
Should be available in [v1.7.0](https://github.com/akopachov/flipper-zero_authenticator/releases/tag/v1.7.0). Keep in mind that by default BadBT is disabled, and you need to [enable it](https://github.com/akopachov/flipper-zero_authenticator/wiki/Is-there-a-token-input-automation%3F) if you would like to use it. Also [there are FAP files](https://github.com/akopachov/flipper-zero_authenticator/wiki/How-to-get-it%3F) where BadBT stuff is physically excluded from the build.

**Reply by ClaraCrazy** [reply]
btw, your list is kinda wrong. if a firmware can run it depends on the api version (well... not for us bc we allow loading apps with outdated ones now, with a warning things may break) but uh... RM, dexv, ours & UL dont share the same api version so your list text is slightly missleading

also, nice work. Funny how everyone wants to use this arguably useless BT feature so much.

**Reply by eried** [reply]
> Should be available in [v1.7.0](https://github.com/akopachov/flipper-zero_authenticator/releases/tag/v1.7.0).

nice thanks! I will try to test soon. 

BTW this is quite nice tool to distribute the faps: 
https://flipc.org/

**Reply by akopachov** [reply]
> btw, your list is kinda wrong. if a firmware can run it depends on the api version (well... not for us bc we allow loading apps with outdated ones now, with a warning things may break) but uh... RM, dexv, ours & UL dont share the same api version so your list text is slightly missleading
> 
> also, nice work. Funny how everyone wants to use this arguably useless BT feature so much.

Good point. Will update that list. Initially I wanted to make it "user friendly" as I doubt lot of users will know what is the API version of a firmware they are using, but I do agree with you, making it friendlier made it incorrect finally.

And yeah, do agree on the arguable BT feature. And that is why I made it easy to build app without this feature at all, which will let me include into all the further releases both versions, with BT and without it. Hope it will make happy everybody who is using this app, regardless of their thoughts on BT feature.

**Reply by eried** [reply]
Hey! :/ I dont think it works on Windows. I see bluetooth doing something but I guess it would require another setting of "delay before sending" so the OS has the keyboard ready. Because when I open the BT app, there is a few seconds of "connecting" in the flipper and then the keyboard works. I think in android this process is fast

**Reply by akopachov** [reply]
Well, it is up to you as a user to wait until BT got paired. Flipper starts advertising via BT as soon as user come to "token generation" screen. Then you may pair it to your device, and once you ensure it is paired and device is ready to accept keyboard commands you may trigger BadBT automation.

**Reply by eried** [reply]
yes, what I mean is BT is off when you are not in the app -> https://photos.app.goo.gl/GYWXfe2wqDFLa1N7A

as seen in the video, so I guess BadBT is not advertising it? (no blue led)

**Reply by eried** [reply]
I see there is something happening (because the hourglass in Windows, so maybe the BLE device is turn on for a brief second, not enough to be registered by the OS as a keyboard, like in BT devices the flipper is just "paired", but it is "connected" when you open the BT keyboard for example, in the other app)

**Reply by akopachov** [reply]
Wait, are you using "arrow down" key on the video? If so, then you should use "arrow up" key instead. As "arrow down" key is reserved for BadUSB automation (just as previously)

**Reply by eried** [reply]
ohhhh! then it works perfectly!! haha lol sorry

**Reply by akopachov** [reply]
Glad it works for you! Thanks for testing it.
